### PR TITLE
Reload only when config changes.

### LINKED ISSUE #2756: Prom2: SIGHUP for config reload causes artifacts in TSDB storage
**What did you do?**
Run P2 from dev-2.0 branch side by side our P1 stack with the same config reloading and scraped targets.

**What did you expect to see?**
A stable retention similarly to P1.

**What did you see instead? Under which circumstances?**

A periodic, every 1h weird whole in our data, which massively crops up in rates.

image

This coincides with our config syncing job that sends a SIGHUP to prometheus at least every hour. The data loss corelates *exactly* with the SIHGUP log statement and with the prometheus_config_last_reload_success_timestamp

image

@fabxc tracked down the issue

### PR DESCRIPTION
**Pull request by Gouthamve**
Pending tests and fixes #2756 

So what happens when we reload config is that we stop all scrape loops (which might have inflight scrapes) and start new ones. This causes some series to "miss" a scrape and causes the artifacts shown in #2756. This could be reproduced in 1.x also.

I propose that we compare the configs and reload only if it changes. But before I write the tests, I want to know if its sane to do this.

cc @brian-brazil @fabxc @beorn7

---
This change is [Reviewable](https://reviewable.io/reviews/prometheus/prometheus/2830)

### CODE REVIEW

**Comment by brian-brazil** [comment]
This looks like it's going to be fragile, as we'll forget to update this when adding new fields. Can we use DeepEquals?

**Comment by Gouthamve** [comment]
So `reflect.DeepEqual` on `ServiceDiscoveryConfigs` fails as StaticConfigs adds a null target. Could not figure out why, but StaticConfig becomes: `[0]` while new config has the field set as `[]`.

**Comment by Gouthamve** [comment]
@brian-brazil The modification is being done by this loop: https://github.com/prometheus/prometheus/blob/master/discovery/discovery.go#L143-L145

We could make a copy of the groups so that the original config is not modified.

**Comment by brian-brazil** [comment]
Yes, in general we should be making copys of everything that's being changed, as that's led to fun bugs in the past.

**Comment by beorn7** [comment]
Anecdotal evidence: The work-around for https://github.com/prometheus/prometheus/issues/2799 is to send a SIGHUP. With this optimization, the work-around doesn't work anymore.

I don't want to say that a work-around for a know bug should keep us from doing the right thing. But this change adds another possibility for bugs (config has changed but Prometheus doesn't recognize that fact). How often does it happen that a SIGHUP is sent _without_ any changes to the config? Or in other words: If the user tells us to reload the config, shouldn't we just trust them and not try to second-guess their intention?

**Comment by fabxc** [comment]
I think that's a very valid point. Who knows what other unintended behavior we have that people happen to not run into because they SIGHUP frequently. Maybe fixing this by not having the context abort in-flight scrapes is the better solution after all?

That is a tricky change though to a code path that  got complicated due to staleness handling. I'd argue that further changes at this point might negatively impact our stabilisation efforts for 2.0. As @Gouthamve  and I both have confirmed now, the disruption reported in #2776 is equally happening in 1.x.

**Comment by Gouthamve** [comment]
@beorn7 The workaround for #2799 will still work. So the part that we are fixing here is this: For the scrapeloop for each target, we are reloading the scrapeloop too cancelling any in-flight scrapes. 

While the user is asking us to reload the config, I am not sure it is fair to cancel scrapes that are untouched by the config change.

**Comment by brian-brazil** [comment]
The other problem is that this wouldn't catch when the content of the bearer token file changes. Personally I'd suggest removing that support, as we don't accept secrets from files anywhere else.

**Comment by brancz** [comment]
> The other problem is that this wouldn't catch when the content of the bearer token file changes. Personally I'd suggest removing that support, as we don't accept secrets from files anywhere else.

There are some problems with this:

1) We default to the file/directory the bearer token file is mounted into by the Kubernetes admission controller, therefore the proposed change would break nearly _every_ existing Prometheus setup on top of Kubernetes.
3) When generating the config in some pipeline, there's no guarantee that pipeline knows the same secrets as the cluster.
4) Using the file is the canonical way of configuring authentication with a Kubernetes API from within a cluster.

I realize the Kubernetes SD is in beta, but this is too much of an invasive change. In this case we just have to deal with the environment, and people running Prometheus on Kubernetes need to find another way of dealing with this problem (eg. a sidecar that reloads on the secret changes).

**Comment by brian-brazil** [comment]
That nixes the idea of removing that then, unless we do something like checking the mtime of the file every time we do a scrape.

**Comment by brian-brazil** [comment]
In relation to bearer token files, I think we should switch to reading them at every scrape. Then a DeepEquals should always work.

**Comment by Gouthamve** [comment]
@brian-brazil @brancz Could you take a look now?

File: retrieval/scrape.go [hunk 1]
```diff
@@ -196,6 +197,9 @@ func (sp *scrapePool) stop() {
 // but all scrape loops are restarted with the new scrape configuration.
```
Comment (brian-brazil) [inline retrieval/scrape.go]: This comment needs updating.

Comment (Gouthamve) [reply retrieval/scrape.go]: Done

File: retrieval/scrape.go [hunk 2]
```diff
@@ -193,9 +194,13 @@ func (sp *scrapePool) stop() {
 }
 
 // reload the scrape pool with the given scrape configuration. The target state is preserved
-// but all scrape loops are restarted with the new scrape configuration.
+// but all scrape loops are restarted with the new scrape configuration when there is a chane in the
```
Comment (brian-brazil) [inline retrieval/scrape.go]: change

**Comment by brancz** [comment]
> In relation to bearer token files, I think we should switch to reading them at every scrape. Then a DeepEquals should always work.

Should this ever become a problem, we can always mmap the file to read it. This is actually a lot more common than it sounds.

The approach and implementation lgtm.

I'd merge, but unfortunately there is a conflict :/

**Comment by krasi-georgiev** [comment]
this needs re-thinking now with the new Discovery refactoring. 

A big flaw with the current design is that when managers are run independently but still depend on the global config at the time or reloading with  `ApplyConfig`  managers are out of sync during the  reloading process 

in other words some managers will run using different configuration which will cause weird results. 
the new discovery manager is affected by this but I imagine some other mangers as well.

https://github.com/prometheus/prometheus/pull/3605
I think I managed to solve the problem in this PR so please give me some initial thoughts.

**Comment by brancz** [comment]
I'm not sure I understand correctly, this PR is about whether a reload happens in the first place.

In the PR you mentioned we still switch/stop/start more or less synchronously, I believe what we actually need is an A/B approach, that atomically switches to a manager instance when necessary reload is required, to properly hand over n reload.

Basically what I'm thinking of is this

* Is reload necessary (the content of this PR)
* Create new managers with new config
* Start new managers (we first must make sure that until they're in place they don't get any work to work on, I'm unaware whether that's the case today)
* Atomically switch out old with new managers essentially activating it, and taking over any "new" work.

**Comment by krasi-georgiev** [comment]
@brancz just spent another 1h comparing the branches and in fact this PR is even obsolete.After the refactoring `func (sp *scrapePool) reload` is never even used (which reminds  me that I need to remove it). 

Instead we just remove or add individual target loops using. 
`func (sp *scrapePool) Sync(tgs []*config.TargetGroup) {`

I asked @Bplotka  to test a binary so hopefully this confirms it.
(it would feel good to kill few issues with one PR  :innocent:)

**Comment by Gouthamve** [comment]
Time for me to read the discovery refactor, but sounds like this is fixed in that :)

**Comment by brian-brazil** [comment]
Wouldn't we at least still need to re-read the bearer token file at every scrape?

**Comment by krasi-georgiev** [comment]
@brian-brazil owe you a 🍺  , completely overlooked this when I did the refactoring.
Here is my suggestion for the fix https://github.com/prometheus/prometheus/pull/3610

@Gouthamve do you want to rebase your PR or would you mind if  I steal your code for the tests and complete my PR?

**Comment by brian-brazil** [comment]
I don't believe that fixes the case of the bearer token changing on disk, as it's not part of the config.

**Comment by krasi-georgiev** [comment]
mh , @brancz suggestion for mmap looks interesting will look into that.

**Comment by krasi-georgiev** [comment]
PR updated with the bearer file read on every scrape and added the test. 
Before the merge it might be best to wait for a confirmation from Bplotka that it fixed the issue for them.

**Pull request closed without merging**